docs(NODE-3753): Update testing readme file (#3048)
mongodb/node-mongodb-native: test/readme.md

check the topology of the MongoDB server that is being used. If the topology does not match, the
tests will be skipped.

Below is a summary of the types of test automation in this repo.

| Type of Test | Test Location | About the Tests | How to Run Tests |
| ------------ | ------------- | --------------- | ---------------- |
| Unit | `/test/unit` | The unit tests test individual pieces of code, typically functions. These tests do **not** interact with a real database, so mocks are used instead. | `npm run check:unit` |
| Functional | `/test/functional` | The functional tests test that a given feature or piece of a feature is working as expected. These tests do **not** use mocks; instead, they interact with a real database. | `npm run check:test` |
| Benchmark | `/test/benchmarks` | The benchmark tests report how long a designated set of tests take to run. They are used to measure performance. | `npm run check:bench` |
| Integration | `/test/integration` | *Coming Soon* The integration tests test that pieces of the driver work together as expected. | `npm run check:test` |

> | Specialized Environment | `/test/manual` | The specalized environment tests are functional tests that require specialized environment setups in Evergreen. <br><br>**Note**: "manual" in the directory path does not refer to tests that should be run manually. These tests are automated. These tests require manual configuration in Evergreen. | There is no single script for running all of the specialized environment tests. Instead, you can run the appropriate script based on the specialized environment you want to use: <br>- `npm run check:atlas` to test Atlas <br>- `npm run check:adl` to test Atlas Data Lake <br>- `npm run check:ocsp` to test OSCP <br>- `npm run check:kerberos` to test Kerberos <br>- `npm run check:tls` to test TLS <br>- `npm run check:ldap` to test LDAP authorization

Reviewer: ```suggestion
| Specialized Environment | `/test/manual` | The specalized environment tests are functional tests that require specialized environment setups in Evergreen. <br><br>**Note**: "manual" in the directory path does not refer to tests that should be run manually. These tests are automated. These tests have a special evergreen configuration and run in isolation from the other tests.. | There is no single script for running all of the specialized environment tests. Instead, you can run the appropriate script based on the specialized environment you want to use: <br>- `npm run check:atlas` to test Atlas <br>- `npm run check:adl` to test Atlas Data Lake <br>- `npm run check:ocsp` to test OSCP <br>- `npm run check:kerberos` to test Kerberos <br>- `npm run check:tls` to test TLS <br>- `npm run check:ldap` to test LDAP authorization
```

> | Spec | Test input and expected results are in `/test/spec`.  <br>Test runners are in `/test/functional` with  the `_spec` suffix in the test file's name.  <br>Some spec tests are also in `/test/unit`. | All of the MongoDB drivers follow the same [specifications (specs)][driver-specs]. Each spec has tests associated with it. Some of the tests are prose (written, descriptive) tests.  Other tests are written in JSON. The developers on the driver teams automate these tests. For prose tests, the tests are converted to automation and stored in the `test/unit` or `test/functional` directory as appropriate. For the JSON tests, a developer uses a tool to convert the JSON test file to YAML. Both the JSON and YAML files are stored in `test/spec`. The test runners for the JSON and YAML files are located in `test/functional` and `/test/unit`. | `npm run check:test` to run all of the functional and integration tests (including the spec tests stored with those). `npm run check:unit` to run all of the unit tests (including the spec tests stored with those).

Reviewer: ```suggestion
| Spec | Test input and expected results are in `/test/spec`.  <br>Test runners are in `/test/functional` with  the `_spec` suffix in the test file's name.  <br>Some spec tests are also in `/test/unit`. | All of the MongoDB drivers follow the same [specifications (specs)][driver-specs]. Each spec has tests associated with it. Some of the tests are prose (written, descriptive) tests.  Other tests are written in YAML and converted to JSON. The developers on the driver teams automate these tests. For prose tests, the tests are converted to automation and stored in the `test/unit` or `test/functional` directory as appropriate. Both the JSON and YAML files are stored in `test/spec`. The test runners for the JSON and YAML files are located in `test/functional` and `/test/unit`. | `npm run check:test` to run all of the functional and integration tests (including the spec tests stored with those). `npm run check:unit` to run all of the unit tests (including the spec tests stored with those).
```

In the following subsections, we'll dig into the details of running the tests.

### Testing Different MongoDB Topologies

As we mentioned earlier, the tests check the topology of the MongoDB server being used and run the tests associated with that topology.  Tests that don't have a matching topology will be skipped.

In the steps above, we started a standalone server: `./test/tools/cluster_setup.sh server`.

You can use the same [cluster_setup.sh](tools/cluster_setup.sh) script to start a replica set or sharded cluster by passing the appropriate option: `./test/tools/cluster_setup.sh replica_set` or
`./test/tools/cluster_setup.sh sharded_cluster`.  If you are running more than a standalone server, make sure your `ulimit` settings are in accordance with [MongoDB's recommendations][mongodb-ulimit]. Changing the settings on the latest versions of macOS can be tricky. See [this article][macos-ulimt] for tips. (You likely don't need to do the complicated maxproc steps.)

The [cluster_setup.sh](tools/cluster_setup.sh) script automatically stores the files associated with the MongoDB server in the `data` directory, which is stored at the top level of this repository.

> You can delete this directory if you want to ensure you're running a clean configuration. If you delete the directory, the associated database server will be stopped, and you will need to run [cluster_setup.sh](tools/cluster_setup.sh) again.

Reviewer: I'm actually not sure if this is true - I think if the server is running you wouldn't be allowed to delete the directory as the server would have a handle on those files. I think you would need to stop the server first, then delete the directory.

Author: This is taken from one of the comments Daria added to https://jira.mongodb.org/browse/NODE-3753. I've tested it out, and the processes are no longer running after deleting the directory. I'm happy to change it if you don't think people should take this approach.

Reviewer: I just verified that as well, so fine to stay as is.

> You can prefix `npm test` with a `MONGODB_URI` environment variable to point the tests to a specific deployment. For example, for a standalone server, you might use: `env MONGODB_URI=mongodb://localhost:27017 npm test`. For a replica set, you might use: `env MONGODB_URI=mongodb://localhost:31000,localhost:31001,localhost:31002/?replicaSet=rs npm test`.

Reviewer: `env` is not necessary here in the command.

Author: Thank you!

### Running Individual Tests

> The easiest way to run a single test is by appending `.only()` to the suite or test you want to run. For example, you could update a test function to be `it.only(‘cool test’, function() {})`.  Then

Reviewer: ```suggestion
The easiest way to run a single test is by appending `.only()` to the test context you want to run. For example, you could update a test function to be `it.only(‘cool test’, function() {})`.  Then
```

run the test using `npm run check:test` for a functional or integration test or `npm run check:unit` for a unit test.  See [Mocha's documentation][mocha-only] for more detailed information on `.only()`.

Another way to run a single test is to use Mocha's `grep` flag.  For functional or integration tests, run `npm run check:test -- -g 'test name'`. For unit tests, run `npm run check:unit -- -g 'test name'`. See the [Mocha documentation][mocha-grep] for information on the `grep` flag.

## Running the Tests in Evergreen

[Evergreen][evergreen-wiki] is the continuous integration (CI) system we use.  Evergreen builds are automatically run whenever a pull request is created or when commits are pushed to particular branches (e.g., main, 4.0, and 3.6).

> Each Evergreen build runs the test suite against a variety of build variants that include a combination of topologies, special environments, and operating systems. By default, commits in pull requests only run a subset of the build variants in order to save time and resources. These builds can be individually configured in the Evergreen UI to include more build variants.

Reviewer: ```suggestion
Each Evergreen build runs the test suite against a variety of build variants that include a combination of topologies, special environments, and operating systems. By default, commits in pull requests only run a subset of the build variants in order to save time and resources. These builds can be individually configured in the `.evergreen/config.yml.in` and then by generating a new Evergreen config via `node .evergreen/generate_evergreen_tasks.js`.
```

##### Setup

Begin by setting up the Evergreen CLI.

1. Download and install the Evergreen CLI according to the instructions in the [Evergreen Documentation][evergreen-docs].
1. Be sure to create `evergreen.yml` as described in the documentation.
1. Add the Evergreen binary to your path.

##### Running the Build

Once you have the Evergreen CLI setup, you are ready to run a build.

1. In a terminal, navigate to your node driver directory:

   `cd node-mongodb-native`
1. Use the Evergreen `patch` command. `-y` skips the confirmation dialog. `-u` includes uncommitted changes. `-p [project name]` specifies the Evergreen project. --browse opens the patch URL in your browser.

   `evergreen patch -y -u -p mongo-node-driver --browse`
1. In your browser, select the build variants and tasks to run.

## Using a Pre-Release Version of a Dependent Library

You may want to test the driver with a pre-release version of a dependent library (e.g., [bson][js-bson]).
Follow the steps below to do so.
1. Open [package.json](../package.json)
1. Identify the line that specifies the dependency

> 1. Replace the version number with the commit hash of the dependent library. For example: `"bson": "e29156f7438fa77c1672fd70789d7ade9ca65061"`

Reviewer: ```suggestion
1. Replace the version number with the commit hash of the dependent library. For example, js-bson project on Github: `"bson": "mongodb/js-bson#e29156f7438fa77c1672fd70789d7ade9ca65061"`
```

1. Run `npm install` to install the dependency

Now you can run the automated tests, run manual tests, or kick off an Evergreen build from your local
repository.

## Manually Testing the Driver

You may want to manually test changes you have made to the driver. The steps below will walk you through how to create a new Node project that uses your local copy of the Node driver.  You can
modify the steps to work with existing Node projects.

1. Navigate to a new directory and create a new Node project by running `npm init` in a terminal and    working through the interactive prompts. A new file named `package.json` will be created for you.
1. In `package.json`, create a new dependency for `mongodb` that points to your local copy of the driver. For example:
   ```
   "dependencies": {
     "mongodb": "/path-to-your-copy-of-the-driver-repo/node-mongodb-native"
   }
   ```
1. Run `npm install` to install the dependency.

> 1. Create a new file that uses the driver to test your changes. See the [MongoDB Node.js Quick Start Repo][node-quick-start] for example scripts you can use.

Reviewer: I'd add something like, "Note when making driver changes you will have to run `npm run build:ts` with each change to have it take effect".

Author: Great catch! Thank you!

In order to test some features, you will need to generate and set a specialized group of environment variables.  The subsections below will walk you through how to generate and set the environment variables for these features.

We recommend using a different terminal for each specialized environment to avoid the environment variables from one specialized environment impacting the test runs for another specialized environment.

Before you begin any of the subsections below, clone the [drivers-evergreen-tools repo](https://github.com/mongodb-labs/drivers-evergreen-tools.git).

We recommend creating an environment variable named `DRIVERS_TOOLS` that stores the path to your local copy of the driver-evergreen-tools repo: `export DRIVERS_TOOLS="/path/to/your/copy/of/drivers-evergreen-tools"`.

### Serverless

The following steps will walk you through how to create and test a MongoDB Serverless instance.

1. Create the following environment variables using a command like `export PROJECT="node-driver"`.

   - `PROJECT`
   - `SERVERLESS_DRIVERS_GROUP`
   - `SERVERLESS_API_PUBLIC_KEY`
   - `SERVERLESS_API_PRIVATE_KEY`
   - `SERVERLESS_ATLAS_USER`
   - `SERVERLESS_ATLAS_PASSWORD`

>    - `LOADBALANCED`

Reviewer: Did you find this environment variable is needed? I didn't see where it was used in the scripts.

Author: For traceability, we tried this out together and decided it wasn't needed.

>    TODO:  Explain what these variables represent and how to get their values.

Reviewer: Maybe add: "If you are an Atlas user the values can be found in your project details. If you are testing for driver work, consult the Evergreen project configuration for these values."

   FAKE_MONGODB_SERVICE_ID="true"
   ```
   > **Note:** `FAKE_MONGODB_SERVICE_ID` will no longer be needed with the completion of [NODE-3431](https://jira.mongodb.org/browse/NODE-3431).
1. Source the environment variables using a command like `source lb.env`.
1. Export **each** of the environment variables that were created in `lb.env`. For example: `export SINGLE_MONGOS_LB_URI`.
1. Run the test suite as you normally would:
   ```sh
   npm run check:test
   ```
   Verify that the output from Mocha includes `[ topology type: load-balanced ]`. This indicates the tests successfully accessed the specialized environment variables for load balancer testing.
1. When you are done testing, shutdown the HAProxy load balancer:
   ```sh
   $DRIVERS_TOOLS/.evergreen/run-load-balancer.sh stop
   ```

### Client-Side Field Level Encryption (CSFLE)

The following steps will walk you through how to run the tests for CSFLE.

1. Install [MongoDB Client Encryption][npm-csfle] if you haven't already:
   `npm install mongodb-client-encryption`

1. Create the following environment variables using a command like `export AWS_REGION="us-east-1"`.

   - `AWS_ACCESS_KEY_ID`
   - `AWS_SECRET_ACCESS_KEY`

>    - `CSFLE_KMS_PROVIDERS`

Reviewer: I'd describe CSFLE_KMS_PROVIDERS as "The raw EJSON description of the KMS providers". The format is:

```
interface CSFLE_kms_providers {
   aws: {
     accessKeyId: string;
     secretAccessKey: string;
  };
  azure: {
    tenantId: string;
    clientId: string;
    clientSecret: string;
  };
  gcp: {
    email: string;
    privateKey: string;
  };
  local: {
    // EJSON handle converting this, its actually the canonical -> { $binary: { base64: string; subType: string } }
    // **NOTE**: The dollar sign has to be escaped when using this as an ENV variable
    key: Binary;
  }
}
```

Author: Than you for this!

   - `AWS_REGION`
   - `AWS_CMK_ID`

>    TODO:  Explain what these variables represent and how to get their values.

Reviewer: Make a note for drivers testing to pull from Evergreen, otherwise pull them from the AWS account.
